gson/src/main/java/com/google/gson/annotations/Required.java

```diff
@@ -0,0 +1,18 @@
+package com.google.gson.annotations;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation causes a {@link com.google.gson.JsonParseException} to be thrown
+ * during deserialization if this field is not present in the input JSON, is null,
+ * or is not of a type assignable to this field.
```

Comment:
> or is not of a type assignable to this field.

That is the default behavior for fields anyways, isn't it? I don't think this is worth mentioning here then, otherwise it might cause confusion whether this does / does not apply to optional fields.

Reply:
Well, if you try to parse something that has an object where you expect a string, the string will be null. The idea behind this annotation is to guarantee that a required field will never, under any circumstances, be null.

Reply:
> Well, if you try to parse something that has an object where you expect a string, the string will be null.

Are you sure? The following throws an exception:
```java
public class GsonParsingTest {
    static class Test {
        String s;
    }
    
    public static void main(String[] args) {
        new Gson().fromJson("{\"s\":{}}", Test.class);
    }
}
```

Maybe there are custom type adapters which return `null` for non-`null` JSON in case of type mismatch, but we cannot generally assume that. Not every type adapter which returns `null` must have encountered wrong JSON data.